File: artifacts/py-api/routes/duplicates.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session
from typing import Optional
import math
from db import get_db, DuplicatePaymentRecord

router = APIRouter()


def to_dict(rec: DuplicatePaymentRecord) -> dict:
    return {
        "id": rec.id,
        "payment1Id": rec.payment1_id,
        "payment2Id": rec.payment2_id,
        "probability": rec.probability,
        "duplicateType": rec.duplicate_type,
        "paymentSystem": rec.payment_system,
        "amount": rec.amount,
        "currency": rec.currency,
        "senderBIC": rec.sender_bic,
        "receiverBIC": rec.receiver_bic,
        "originatorCountry": rec.originator_country,
        "beneficiaryCountry": rec.beneficiary_country,
        "paymentDate1": rec.payment_date1,
        "paymentDate2": rec.payment_date2,
        "status": rec.status,
        "matchedFields": rec.matched_fields or [],
        "detectedAt": rec.detected_at.isoformat() if rec.detected_at else None,
        "notes": rec.notes,
    }


@router.get("")
def get_duplicate_payments(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=50, ge=1, le=500),
    minProbability: Optional[float] = None,
    paymentSystem: Optional[str] = None,
    status: Optional[str] = None,
    dateFrom: Optional[str] = None,
    dateTo: Optional[str] = None,
    db: Session = Depends(get_db),
):
    query = db.query(DuplicatePaymentRecord)

    if minProbability is not None:
        query = query.filter(DuplicatePaymentRecord.probability >= minProbability)
    if paymentSystem:
        query = query.filter(DuplicatePaymentRecord.payment_system == paymentSystem)
    if status:
        query = query.filter(DuplicatePaymentRecord.status == status)

    query = query.order_by(DuplicatePaymentRecord.probability.desc(), DuplicatePaymentRecord.detected_at.desc())
    
    total = query.count()
    offset = (page - 1) * limit
    items = query.offset(offset).limit(limit).all()
    
    return {
        "items": [to_dict(r) for r in items],
        "total": total,
        "page": page,
        "limit": limit,
        "totalPages": math.ceil(total / limit) if total > 0 else 1,
    }


@router.get("/{dup_id}")
def get_duplicate_by_id(dup_id: str, db: Session = Depends(get_db)):
    rec = db.query(DuplicatePaymentRecord).filter(DuplicatePaymentRecord.id == dup_id).first()
    if not rec:
        raise HTTPException(status_code=404, detail="Duplicate payment record not found")
    
    result = to_dict(rec)
    result["payment1Details"] = {
        "id": rec.payment1_id,
        "paymentSystem": rec.payment_system,
        "amount": rec.amount,
        "currency": rec.currency,
        "senderBIC": rec.sender_bic,
        "receiverBIC": rec.receiver_bic,
        "valueDate": rec.payment_date1,
        "originatorCountry": rec.originator_country,
    }
    result["payment2Details"] = {
        "id": rec.payment2_id,
        "paymentSystem": rec.payment_system,
        "amount": rec.amount,
        "currency": rec.currency,
        "senderBIC": rec.sender_bic,
        "receiverBIC": rec.receiver_bic,
        "valueDate": rec.payment_date2,
        "beneficiaryCountry": rec.beneficiary_country,
    }
    result["agentExplanations"] = []
    return result


@router.patch("/{dup_id}/status")
def update_duplicate_status(
    dup_id: str,
    body: dict,
    db: Session = Depends(get_db),
):
    rec = db.query(DuplicatePaymentRecord).filter(DuplicatePaymentRecord.id == dup_id).first()
    if not rec:
        raise HTTPException(status_code=404, detail="Record not found")
    
    valid_statuses = ["pending", "confirmed_duplicate", "dismissed", "under_review"]
    new_status = body.get("status")
    if new_status not in valid_statuses:
        raise HTTPException(status_code=400, detail=f"Invalid status. Must be one of: {valid_statuses}")
    
    rec.status = new_status
    if "notes" in body:
        rec.notes = body["notes"]
    
    db.commit()
    db.refresh(rec)
    return to_dict(rec)

File: artifacts/api-server/src/py-api/routes/duplicates.ts
import { Router } from "express";
import { query, queryOne, execute } from "../db.js";
import { v4 as uuidv4 } from "uuid";

const router = Router();

function toCamel(r: Record<string, unknown>) {
  return {
    id: r["id"],
    payment1Id: r["payment1_id"],
    payment2Id: r["payment2_id"],
    probability: Number(r["probability"] ?? 0),
    duplicateType: r["duplicate_type"],
    paymentSystem: r["payment_system"],
    amount: Number(r["amount"] ?? 0),
    currency: r["currency"],
    senderBIC: r["sender_bic"],
    receiverBIC: r["receiver_bic"],
    originatorCountry: r["originator_country"],
    beneficiaryCountry: r["beneficiary_country"],
    paymentDate1: r["payment_date1"],
    paymentDate2: r["payment_date2"],
    status: r["status"],
    matchedFields: r["matched_fields"],
    detectedAt: r["detected_at"],
    notes: r["notes"],
    scanId: r["scan_id"],
  };
}

router.get("/", async (req, res) => {
  try {
    const page = Math.max(1, Number(req.query["page"] || 1));
    const pageSize = Math.min(200, Math.max(1, Number(req.query["page_size"] || 50)));
    const offset = (page - 1) * pageSize;

    const conditions: string[] = [];
    const params: unknown[] = [];
    let pi = 1;

    const { payment_system, status, min_probability, max_probability, search } = req.query as Record<string, string>;

    if (payment_system) { conditions.push(`payment_system = $${pi++}`); params.push(payment_system); }
    if (status) { conditions.push(`status = $${pi++}`); params.push(status); }
    if (min_probability) { conditions.push(`probability >= $${pi++}`); params.push(Number(min_probability)); }
    if (max_probability) { conditions.push(`probability <= $${pi++}`); params.push(Number(max_probability)); }
    if (search) {
      conditions.push(`(payment1_id ILIKE $${pi} OR payment2_id ILIKE $${pi} OR sender_bic ILIKE $${pi} OR receiver_bic ILIKE $${pi})`);
      params.push(`%${search}%`); pi++;
    }

    const where = conditions.length ? `WHERE ${conditions.join(" AND ")}` : "";

    const [countRow] = await query<{ total: string }>(
      `SELECT COUNT(*) AS total FROM dup_duplicate_payments ${where}`,
      params
    );
    const total = Number(countRow.total);

    const rows = await query(
      `SELECT * FROM dup_duplicate_payments ${where} ORDER BY probability DESC, detected_at DESC LIMIT $${pi} OFFSET $${pi + 1}`,
      [...params, pageSize, offset]
    );

    res.json({
      total,
      page,
      pageSize,
      pages: Math.max(1, Math.ceil(total / pageSize)),
      duplicates: rows.map(toCamel),
    });
  } catch (e) {
    console.error("List duplicates error:", e);
    res.status(500).json({ error: "Failed to list duplicates" });
  }
});

router.get("/:id", async (req, res) => {
  try {
    const row = await queryOne(`SELECT * FROM dup_duplicate_payments WHERE id = $1`, [req.params["id"]]);
    if (!row) return res.status(404).json({ error: "Not found" });
    return res.json(toCamel(row));
  } catch (e) {
    console.error("Get duplicate error:", e);
    return res.status(500).json({ error: "Failed to get duplicate" });
  }
});

router.patch("/:id/status", async (req, res) => {
  try {
    const { status, notes } = req.body as { status: string; notes?: string };
    const validStatuses = ["pending", "confirmed_duplicate", "dismissed", "under_review"];
    if (!validStatuses.includes(status)) {
      return res.status(400).json({ error: "Invalid status" });
    }

    const params: unknown[] = [status, req.params["id"]];
    let sql = `UPDATE dup_duplicate_payments SET status = $1`;
    if (notes !== undefined) { sql += `, notes = $3`; params.push(notes); }
    sql += ` WHERE id = $2 RETURNING *`;

    const rows = await query(sql, params);
    if (!rows.length) return res.status(404).json({ error: "Not found" });
    return res.json(toCamel(rows[0]));
  } catch (e) {
    console.error("Update status error:", e);
    return res.status(500).json({ error: "Failed to update status" });
  }
});

router.delete("/:id", async (req, res) => {
  try {
    const count = await execute(`DELETE FROM dup_duplicate_payments WHERE id = $1`, [req.params["id"]]);
    if (!count) return res.status(404).json({ error: "Not found" });
    return res.json({ success: true });
  } catch (e) {
    console.error("Delete duplicate error:", e);
    return res.status(500).json({ error: "Failed to delete" });
  }
});

export default router;
